packages/layout/src/Modal.ts

```diff
@@ -0,0 +1,204 @@
+import { d3SelectionType, HTMLWidget, publish, publishProxy, Widget } from "@hpcc-js/common";
+
+import "../src/Modal.css";
+
+export class Modal extends HTMLWidget {
+
+    private _widget: Widget;
+
+    private _relativeTarget: HTMLElement;
+
+    private _fade: d3SelectionType;
+    private _modal: d3SelectionType;
+    private _modalHeader: d3SelectionType;
+    private _modalBody: d3SelectionType;
+    private _close: any;
+
+    constructor() {
+        super();
+        this._tag = "div";
+    }
+
+    closeModal() {
+        if (typeof this._close === "function") {
+            this._close();
+        }
+        this.show(false).render();
+    }
+
+    getRelativeTarget() {
+        let relativeTarget;
+        if (this.relativeTargetId()) {
+            relativeTarget = document.getElementById(this.relativeTargetId());
+            if (relativeTarget) {
+                return relativeTarget;
+            }
+        }
+        if (!relativeTarget) {
+            relativeTarget = this.locateAncestor("layout_Grid");
+            if (relativeTarget && relativeTarget.element) {
+                return relativeTarget.element().node();
+            }
+        }
+        return document.body;
+    }
+
+    setModalSizeLimits() {
+        if (this.minHeight() || this.minWidth()) {
+            this._modal
+                .style("min-height", this.minHeight())
+                .style("min-width", this.minWidth())
+                .style("max-height", this.maxHeight())
+                .style("max-width", this.maxWidth())
+                ;
+        }
+    }
+
+    setFadePosition(rect) {
+        this._fade.node().style.top = rect.top + "px";
+        this._fade.node().style.left = rect.left + "px";
+        this._fade.node().style.width = rect.width + "px";
+        this._fade.node().style.height = rect.height + "px";
+    }
+
+    setModalPosition(rect) {
+        const _node = this._modal.node();
+        const contentRect = _node.getBoundingClientRect();
+        _node.style.position = "fixed";
+        _node.style.top = (rect.top + (rect.height / 2) - (contentRect.height / 2)) + "px";
+        _node.style.left = (rect.left + (rect.width / 2) - (contentRect.width / 2)) + "px";
+        _node.style.width = contentRect.width + "px";
+        _node.style.height = contentRect.height + "px";
+        if (this.minHeight() || this.minWidth()) {
+            _node.style["min-height"] = this.minHeight();
+            _node.style["min-width"] = this.minWidth();
+        }
+    }
+
+    enter(domNode, element) {
+        super.enter(domNode, element);
+        const context = this;
+        this._fade = element.append("div")
+            .classed("layout_Modal-fade", true)
```

Comment:
It should not need the "layout_Modal" prefix as that is already assigned to a parent div, so out selectors can still be specific: `.layout_Modal > div.fade` type of thing?

Reply:
`.fade` is too common of a name. Could cause issues.